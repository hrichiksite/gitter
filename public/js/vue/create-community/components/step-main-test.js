const mount = require('../../__test__/vuex-mount');
const { default: StepMain } = require('./step-main.vue');

const {
  createOrgGitlabGroupFixture,
  createRepoGithubRepoFixture
} = require('../../__test__/fixture-helpers');

import {
  CREATE_COMMUNITY_STEP_BACKING_ENTITY_GITLAB,
  CREATE_COMMUNITY_STEP_BACKING_ENTITY_GITHUB
} from '../constants';

describe('StepMain', () => {
  it('GitLab user matches snapshot', () => {
    const { wrapper } = mount(StepMain, {}, store => {
      store.state.user = {
        username: 'some-username_gitlab',
        providers: ['gitlab']
      };
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('GitHub user matches snapshot', () => {
    const { wrapper } = mount(StepMain, {}, store => {
      store.state.user = {
        username: 'some-username',
        providers: ['github']
      };
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('filling out the name/slug', () => {
    it('should update community name when user adds input', () => {
      const { wrapper, stubbedActions } = mount(StepMain, {});
      wrapper.find({ ref: 'communityNameInput' }).element.value = 'hello';
      wrapper.find({ ref: 'communityNameInput' }).trigger('input');

      expect(stubbedActions.createCommunity.setCommunityName).toHaveBeenCalledWith(
        expect.anything(),
        'hello',
        undefined
      );
    });

    it('should update community slug when user adds input', () => {
      const { wrapper, stubbedActions } = mount(StepMain, {});
      wrapper.find({ ref: 'communitySlugInput' }).element.value = 'some-slug';
      wrapper.find({ ref: 'communitySlugInput' }).trigger('input');

      expect(stubbedActions.createCommunity.updateCommunitySlug).toHaveBeenCalledWith(
        expect.anything(),
        'some-slug',
        undefined
      );
    });
  });

  describe('errors', () => {
    it('community name error', () => {
      const { wrapper } = mount(StepMain, {}, store => {
        store.state.createCommunity.communityNameError = 'SOME BAD ERROR!';
      });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('community slug error', () => {
      const { wrapper } = mount(StepMain, {}, store => {
        store.state.createCommunity.communitySlugError = 'SOME BAD ERROR!';
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('prompt links', () => {
    it('clicking GitLab prompt moves to backing entity GitLab step', () => {
      const { wrapper, stubbedActions } = mount(StepMain, {}, store => {
        store.state.user = {
          username: 'some-username_gitlab',
          providers: ['gitlab']
        };
      });

      wrapper.find({ ref: 'backingEntityPromptGitlabLink' }).trigger('click');

      expect(stubbedActions.createCommunity.moveToStep).toHaveBeenCalledWith(
        expect.anything(),
        CREATE_COMMUNITY_STEP_BACKING_ENTITY_GITLAB,
        undefined
      );
    });

    it('clicking GitHub prompt moves to backing entity GitHub step', () => {
      const { wrapper, stubbedActions } = mount(StepMain, {}, store => {
        store.state.user = {
          username: 'some-username',
          providers: ['github']
        };
      });

      wrapper.find({ ref: 'backingEntityPromptGithubLink' }).trigger('click');

      expect(stubbedActions.createCommunity.moveToStep).toHaveBeenCalledWith(
        expect.anything(),
        CREATE_COMMUNITY_STEP_BACKING_ENTITY_GITHUB,
        undefined
      );
    });
  });

  describe('entity selected', () => {
    it('matches snapshot', () => {
      const { wrapper } = mount(StepMain, {}, store => {
        store.state.user = {
          username: 'some-username_gitlab',
          providers: ['gitlab']
        };

        store.state.createCommunity.selectedBackingEntity = createOrgGitlabGroupFixture(
          'gitlab-org/gitter'
        );
      });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('clicking change moves back to backing entity step', () => {
      const { wrapper, stubbedActions } = mount(StepMain, {}, store => {
        store.state.user = {
          username: 'some-username_gitlab',
          providers: ['gitlab']
        };

        store.state.createCommunity.selectedBackingEntity = createOrgGitlabGroupFixture(
          'gitlab-org/gitter'
        );
      });

      wrapper.find({ ref: 'changeBackingEntityLink' }).trigger('click');

      expect(stubbedActions.createCommunity.moveToStep).toHaveBeenCalledWith(
        expect.anything(),
        CREATE_COMMUNITY_STEP_BACKING_ENTITY_GITLAB,
        undefined
      );
    });

    it('clicking clear x button clears selected backing entity', () => {
      const { wrapper, stubbedActions } = mount(StepMain, {}, store => {
        store.state.createCommunity.selectedBackingEntity = createOrgGitlabGroupFixture(
          'gitlab-org/gitter'
        );
      });

      wrapper.find({ ref: 'clearAssociationButton' }).trigger('click');

      expect(stubbedActions.createCommunity.clearBackingEntity).toHaveBeenCalledWith(
        expect.anything(),
        undefined,
        undefined
      );
    });

    it('clicking badger checkbox changes allowBadger', () => {
      const { wrapper, stubbedActions } = mount(StepMain, {}, store => {
        store.state.user = {
          username: 'some-username',
          providers: ['github']
        };

        store.state.createCommunity.allowBadger = true;
        store.state.createCommunity.selectedBackingEntity = createRepoGithubRepoFixture(
          'gitterhq/gitter'
        );
      });

      wrapper.find({ ref: 'allowBadgerCheckbox' }).trigger('click');

      expect(stubbedActions.createCommunity.setAllowBadger).toHaveBeenCalledWith(
        expect.anything(),
        false,
        undefined
      );
    });
  });

  describe('submit button', () => {
    it('submit loading user matches snapshot', () => {
      const { wrapper } = mount(StepMain, {}, store => {
        store.state.createCommunity.communitySubmitRequest.loading = true;
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it('submit error user matches snapshot', () => {
      const { wrapper } = mount(StepMain, {}, store => {
        store.state.createCommunity.communitySubmitRequest.error = new Error(
          'My fake request failed'
        );
      });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('clicking submit', () => {
      const { wrapper, stubbedActions } = mount(StepMain, {});
      wrapper.find({ ref: 'submitButton' }).trigger('click');

      expect(stubbedActions.createCommunity.submitCommunity).toHaveBeenCalledWith(
        expect.anything(),
        undefined,
        undefined
      );
    });
  });
});
