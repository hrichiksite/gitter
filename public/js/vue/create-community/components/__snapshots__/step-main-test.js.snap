// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepMain GitHub user matches snapshot 1`] = `
<div>
  <h1
    class="primary-community-heading"
  >
    
    Create community
  
  </h1>
   
  <h2
    class="secondary-community-heading"
  >
    
    Think big,
    <br />
    
    start small, grow fast
  
  </h2>
   
  <input
    class="primary-community-name-input"
    maxlength="80"
    pattern=".+"
    placeholder="your community name"
    required="required"
  />
   
  <!---->
   
  <div
    class="slug-input-wrapper"
  >
    <span
      class="slug-input-prefix"
    >
      gitter.im/
    </span>
     
    <input
      class="slug-input"
      maxlength="80"
      pattern="[a-zA-Z0-9\\\\-_]+"
      placeholder="your-community-uri"
      required="required"
    />
  </div>
   
  <!---->
   
  <!---->
   
  <p
    class="backing-entity-prompt-copy"
  >
    
    do you want to start a community for one of your
    
    <a
      data-disable-routing="1"
      href="#0"
    >
      
      GitHub orgs/repos?
    
    </a>
  </p>
   
  <p>
    
    do you want to start a communtiy for your
    
    <a
      data-disable-routing="1"
      href="#0"
    >
      GitHub
      @some-username user
    
    </a>
  </p>
   
  <!---->
   
  <div
    class="submit-button-container"
  >
    <!---->
     
    <button
      class="submit-button button-caribbean"
      type="submit"
    >
      
      Submit
    
    </button>
  </div>
</div>
`;

exports[`StepMain GitLab user matches snapshot 1`] = `
<div>
  <h1
    class="primary-community-heading"
  >
    
    Create community
  
  </h1>
   
  <h2
    class="secondary-community-heading"
  >
    
    Think big,
    <br />
    
    start small, grow fast
  
  </h2>
   
  <input
    class="primary-community-name-input"
    maxlength="80"
    pattern=".+"
    placeholder="your community name"
    required="required"
  />
   
  <!---->
   
  <div
    class="slug-input-wrapper"
  >
    <span
      class="slug-input-prefix"
    >
      gitter.im/
    </span>
     
    <input
      class="slug-input"
      maxlength="80"
      pattern="[a-zA-Z0-9\\\\-_]+"
      placeholder="your-community-uri"
      required="required"
    />
  </div>
   
  <!---->
   
  <p
    class="backing-entity-prompt-copy"
  >
    
    do you want to start a community for one of your
    
    <a
      data-disable-routing="1"
      href="#0"
    >
      
      GitLab groups/projects?
    
    </a>
  </p>
   
  <!---->
   
  <p>
    
    do you want to start a communtiy for your
    
    <a
      data-disable-routing="1"
      href="#0"
    >
      GitLab
      @some-username user
    
    </a>
  </p>
   
  <!---->
   
  <div
    class="submit-button-container"
  >
    <!---->
     
    <button
      class="submit-button button-caribbean"
      type="submit"
    >
      
      Submit
    
    </button>
  </div>
</div>
`;

exports[`StepMain entity selected matches snapshot 1`] = `
<div>
  <h1
    class="primary-community-heading"
  >
    
    Create community
  
  </h1>
   
  <h2
    class="secondary-community-heading"
  >
    
    Think big,
    <br />
    
    start small, grow fast
  
  </h2>
   
  <input
    class="primary-community-name-input"
    maxlength="80"
    pattern=".+"
    placeholder="your community name"
    required="required"
  />
   
  <!---->
   
  <div
    class="slug-input-wrapper"
  >
    <span
      class="slug-input-prefix"
    >
      gitter.im/
    </span>
     
    <input
      class="slug-input"
      maxlength="80"
      pattern="[a-zA-Z0-9\\\\-_]+"
      placeholder="your-community-uri"
      required="required"
    />
  </div>
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="backing-entity-copy"
  >
    
    Associated with
    
    <div
      class="backing-entity-copy-main"
    >
      <a
        data-disable-routing="1"
        href="https://gitlab.com/groups/gitlab-org/gitter"
        target="_blank"
      >
        <!---->
         
        <i
          class="icon-gitlab"
        />
         
        <span>
          gitlab-org/gitter
        </span>
      </a>
       
      <button
        class="clear-association-button"
      >
        <span
          class="clear-association-icon"
        />
      </button>
       
      <!---->
    </div>
     
    <button
      class="secondary-button-caribbean"
    >
      
      change
    
    </button>
  </div>
   
  <div
    class="submit-button-container"
  >
    <!---->
     
    <button
      class="submit-button button-caribbean"
      type="submit"
    >
      
      Submit
    
    </button>
  </div>
</div>
`;

exports[`StepMain errors community name error 1`] = `
<div>
  <h1
    class="primary-community-heading"
  >
    
    Create community
  
  </h1>
   
  <h2
    class="secondary-community-heading"
  >
    
    Think big,
    <br />
    
    start small, grow fast
  
  </h2>
   
  <input
    class="primary-community-name-input"
    maxlength="80"
    pattern=".+"
    placeholder="your community name"
    required="required"
  />
   
  <div
    class="error-text error-box"
  >
    
    SOME BAD ERROR!
  
  </div>
   
  <div
    class="slug-input-wrapper"
  >
    <span
      class="slug-input-prefix"
    >
      gitter.im/
    </span>
     
    <input
      class="slug-input"
      maxlength="80"
      pattern="[a-zA-Z0-9\\\\-_]+"
      placeholder="your-community-uri"
      required="required"
    />
  </div>
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="submit-button-container"
  >
    <!---->
     
    <button
      class="submit-button button-caribbean"
      type="submit"
    >
      
      Submit
    
    </button>
  </div>
</div>
`;

exports[`StepMain errors community slug error 1`] = `
<div>
  <h1
    class="primary-community-heading"
  >
    
    Create community
  
  </h1>
   
  <h2
    class="secondary-community-heading"
  >
    
    Think big,
    <br />
    
    start small, grow fast
  
  </h2>
   
  <input
    class="primary-community-name-input"
    maxlength="80"
    pattern=".+"
    placeholder="your community name"
    required="required"
  />
   
  <!---->
   
  <div
    class="slug-input-wrapper"
  >
    <span
      class="slug-input-prefix"
    >
      gitter.im/
    </span>
     
    <input
      class="slug-input"
      maxlength="80"
      pattern="[a-zA-Z0-9\\\\-_]+"
      placeholder="your-community-uri"
      required="required"
    />
  </div>
   
  <div
    class="error-text error-box"
  >
    
    SOME BAD ERROR!
  
  </div>
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="submit-button-container"
  >
    <!---->
     
    <button
      class="submit-button button-caribbean"
      type="submit"
    >
      
      Submit
    
    </button>
  </div>
</div>
`;

exports[`StepMain submit button submit error user matches snapshot 1`] = `
<div>
  <h1
    class="primary-community-heading"
  >
    
    Create community
  
  </h1>
   
  <h2
    class="secondary-community-heading"
  >
    
    Think big,
    <br />
    
    start small, grow fast
  
  </h2>
   
  <input
    class="primary-community-name-input"
    maxlength="80"
    pattern=".+"
    placeholder="your community name"
    required="required"
  />
   
  <!---->
   
  <div
    class="slug-input-wrapper"
  >
    <span
      class="slug-input-prefix"
    >
      gitter.im/
    </span>
     
    <input
      class="slug-input"
      maxlength="80"
      pattern="[a-zA-Z0-9\\\\-_]+"
      placeholder="your-community-uri"
      required="required"
    />
  </div>
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="submit-button-container"
  >
    <div
      class="error-text error-box"
    >
      
      Error submitting community: Error: My fake request failed.
    
    </div>
     
    <button
      class="submit-button button-caribbean"
      type="submit"
    >
      
      Submit
    
    </button>
  </div>
</div>
`;

exports[`StepMain submit button submit loading user matches snapshot 1`] = `
<div>
  <h1
    class="primary-community-heading"
  >
    
    Create community
  
  </h1>
   
  <h2
    class="secondary-community-heading"
  >
    
    Think big,
    <br />
    
    start small, grow fast
  
  </h2>
   
  <input
    class="primary-community-name-input"
    maxlength="80"
    pattern=".+"
    placeholder="your community name"
    required="required"
  />
   
  <!---->
   
  <div
    class="slug-input-wrapper"
  >
    <span
      class="slug-input-prefix"
    >
      gitter.im/
    </span>
     
    <input
      class="slug-input"
      maxlength="80"
      pattern="[a-zA-Z0-9\\\\-_]+"
      placeholder="your-community-uri"
      required="required"
    />
  </div>
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="submit-button-container"
  >
    <loading-spinner-stub />
     
    <button
      class="submit-button button-caribbean"
      type="submit"
    >
      
      Submit
    
    </button>
  </div>
</div>
`;
